engine/src/cubos/engine/imgui/data_inspector.cpp

```diff
@@ -0,0 +1,309 @@
+            }
+        });
+    }
+    else if (type.has<FieldsTrait>())
+    {
+        const auto& trait = type.get<FieldsTrait>();
+        const auto& view = trait.view(value);
+        if (trait.size() == 1)
+        {
+            auto [field, fieldVal] = *view.begin();
+            changed |= inspect(name, field->type(), fieldVal, readOnly);
+        }
+        else
+        {
+            showStructured(name, type, [&]() {
+                for (auto [field, fieldVal] : view)
+                {
+                    changed |= inspect(field->name(), field->type(), fieldVal, readOnly);
+                }
+            });
+        }
+    }
+    else
+    {
+        if (readOnly)
+        {
+            showPrimitiveObject(name, type, value);
+        }
+        else
+        {
+            changed |= editPrimitive(name, type, value);
+        }
+    }
+
+    return changed;
+}
+
+/// @brief Returns a string represeting primitive value, depending on it's [TYPE].
+#define RETURN_FORMAT_STRING_IF_TYPE(TYPE)                                                                             \
+    if (type.is<TYPE>())                                                                                               \
+    {                                                                                                                  \
+        return std::to_string(*(TYPE*)(value));                                                                        \
+    }
+
+std::string DataInspector::getPrimitiveValue(const void* value, const Type& type)
+{
+    RETURN_FORMAT_STRING_IF_TYPE(int8_t)
+    RETURN_FORMAT_STRING_IF_TYPE(int8_t)
+    RETURN_FORMAT_STRING_IF_TYPE(int32_t)
+    RETURN_FORMAT_STRING_IF_TYPE(uint8_t)
+    RETURN_FORMAT_STRING_IF_TYPE(uint16_t)
+    RETURN_FORMAT_STRING_IF_TYPE(uint32_t)
+    RETURN_FORMAT_STRING_IF_TYPE(uint64_t)
+    RETURN_FORMAT_STRING_IF_TYPE(float)
+    RETURN_FORMAT_STRING_IF_TYPE(double)
+    if (type.is<bool>())
+    {
+        return *(bool*)(value) ? "true" : "false";
+    }
+    if (type.is<std::string>())
+    {
+        mTextBuffer = *static_cast<const std::string*>(value);
+        return mTextBuffer;
+    }
+
+    CUBOS_ERROR("Could not get primitive object as formatted string (unknown primitive).");
+    return {"???"};
+}
+
+void DataInspector::showPrimitiveObject(const std::string& name, const Type& type, const void* value)
+{
+    ImGui::TableSetColumnIndex(0);
+    ImGui::AlignTextToFramePadding();
+    ImGui::TreeNodeEx(name.c_str(),
+                      ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen | ImGuiTreeNodeFlags_Bullet);
+    if (ImGui::IsItemHovered(ImGuiHoveredFlags_AllowWhenDisabled))
+    {
+        ImGui::SetTooltip("%s", type.name().c_str());
+    }
+
+    std::string str = getPrimitiveValue(value, type);
+    ImGui::TableSetColumnIndex(1);
+    ImGui::TextUnformatted(str.c_str());
+}
+
+void DataInspector::showStructured(const std::string& name, const Type& type, auto fn)
```

Comment:
Maybe rename this method to `withStructure` or something since it is both used 'edit' and 'show' methods.